yb-voyager/src/query/queryissue/parser_issue_detector.go

```diff
@@ -994,63 +1007,145 @@ func (p *ParserIssueDetector) DetectMissingForeignKeyIndexes() []QueryIssue {
+	var options [][]string
+	// Currently we only detect PK recommendations for UNIQUE constraints
+	// TODO: We should also consider UNIQUE INDEXES for PK recommendations
+	// https://yugabyte.atlassian.net/browse/DB-18078
+	for _, uniqueCols := range tm.GetUniqueConstraints() {
+		// Check if all unique columns are NOT NULL
+		allNN := true
+		for _, col := range uniqueCols {
+			if !tm.IsColumnNotNull(col) {
+				allNN = false
+				break
+			}
 		}
 
-		// Collect all qualifying UNIQUE column sets where all columns are NOT NULL
-		var options [][]string
-		// Currently we only detect PK recommendations for UNIQUE constraints
-		// TODO: We should also consider UNIQUE INDEXES for PK recommendations
-		// https://yugabyte.atlassian.net/browse/DB-18078
-		for _, uniqueCols := range tm.GetUniqueConstraints() {
-			allNN := true
-			for _, col := range uniqueCols {
-				if !tm.IsColumnNotNull(col) {
-					allNN = false
-					break
+		if allNN {
+			// For partitioned tables, ensure partition columns are included as the PKs must include
+			// all partition columns down the entire hierarchy
+			if isPartitioned {
+				missingPartitionCols, _ := lo.Difference(partitionColumns, uniqueCols)
+				if len(missingPartitionCols) > 0 {
+					continue
 				}
 			}
-			if allNN {
-				// Create fully qualified column names for this option
-				qualifiedCols := make([]string, len(uniqueCols))
-				for i, col := range uniqueCols {
-					qualifiedCols[i] = fmt.Sprintf("%s.%s", tableName, col)
-				}
-				options = append(options, qualifiedCols)
+
+			// Create fully qualified column names for this option
+			qualifiedCols := make([]string, len(uniqueCols))
+			for i, col := range uniqueCols {
+				qualifiedCols[i] = fmt.Sprintf("%s.%s", tm.GetObjectName(), col)
 			}
-		}
-		if len(options) > 0 {
-			issues = append(issues, NewMissingPrimaryKeyWhenUniqueNotNullIssue("TABLE", tableName, options))
+			options = append(options, qualifiedCols)
 		}
 	}
+
+	if len(options) > 0 {
+		issues = append(issues, NewMissingPrimaryKeyWhenUniqueNotNullIssue("TABLE", tm.GetObjectName(), options))
+	}
+
 	return issues
 }
 
+// getAllPartitionColumnsInHierarchy returns all partition columns down the entire hierarchy
+// for a root-level partitioned table. This includes partition columns from all child tables.
+func (p *ParserIssueDetector) getAllPartitionColumnsInHierarchy(tm *TableMetadata) []string {
+	// Using a map to do automatic deduplication in case if same column is present in multiple child tables
+	allPartitionColumns := make(map[string]bool)
+
+	for _, col := range tm.GetPartitionColumns() {
+		allPartitionColumns[col] = true
+	}
+
+	tableNameOnly := tm.TableName
+	for _, childTM := range p.tablesMetadata {
+		if childTM.PartitionedFrom == tableNameOnly {
+			// Recursively get partition columns from this child and all its descendants
+			childPartitionColumns := p.getAllPartitionColumnsInHierarchy(childTM)
+			for _, col := range childPartitionColumns {
```

Comment:
instead of having this recursive function for multiple purposes like getting columns and checking PK etc.. can we have one more field in `TablesMetadata` as `Partitions []TableMetadata` which will be the list of partitions of that table and this we can populate at once during finalization of this metadata. And then we can utilize that information directly everywhere instead of going and checking all tables again